### DISCUSSION TITLE
Node-locked offline license files with docker

### DISCUSSION THREAD

**Discussion by al-dpopowich** [keygen-sh/community#157] [General]
Preface: I know the best we can do in offline licensing is obfuscate, especially for interpreted languages like python; who can stop a customer from tweaking the code to skip license validation altogether?  Solution: We trust our customers.  Have them sign contracts.  We make it hard to subvert the process.  We obfuscate.  That said...

I have been writing/supporting a webapp software suite that my customer hosts in the cloud.  They sell subscriptions to their customers for use of the webapp. But they have a few customers who will not use their software in the cloud and demand a behind-the-fence solution.  They want to ship docker images and have them (ideally) node-locked.  This is similar to #149, and as discussed there, there's no facility keygen can provide to prevent the 3rd party from copying the distribution and license-file to other hosts.  Add to the mix that you cannot ID a docker container.

So here's an idea I'd like feedback on:

Let's say my customer is Acme and their customer is Customer...

1. Acme ships distribution to Customer with no license-file.
2. Customer runs script to generate an `activation-request`.  This process:
   1. generate RSA key pair
   2. generate random 64 bytes (e.g., `os.urandom(64)`)
   3. store the private-key and randomness in a very obfuscated way.  {{hand waving here}}
   4. using private-key, sign the randomness as the `activation-request`.
   5. the `activation-request` and public-key are packaged and Customer sends it back to Acme.
3. Acme stores `activation-request` along with public-key securely.
4. Acme runs script which verifies signature of `activation-request` then using the public-key, encrypts the `activation-request`.
5. Acme generates a license for Customer and in the metadata places:
```
   machine: {encrypted-activation-request}
```
6. Acme does a checkout of the license and delivers the generated license-file to Customer.
7. Customer installs license-file.
8. When distro runs, in addition to normal validation of license, it will look at the metadata to complete validation:
   1. `machine` must be present.
   2. the private-key must be able to successfully decrypt the value.
   3. the decrypted value must match the original randomness.
 
Yes, we're still obfuscating, but depending on the sophistication therein, we can make it really, really hard for a single customer to copy the distro to their subsidiaries' sites.

Thoughts on this scheme?

**Comment by ezekg** [comment]
We've been working on a product for this exact use-case: https://github.com/keygen-sh/keygen-relay [(spec)](https://github.com/keygen-sh/keygen-relay/blob/master/SPEC.md)

Still a WIP, but at the very least a beta should be out soon.

**Reply by al-dpopowich** [reply]
That looks intriguing! 

Sometimes when I read your documentation I get turned around by the use of "client" and "server", so I'm a bit confused by this:

> Relay will have 2 parts:
> 
> 1. a vendor-facing CLI; and
> 2. an app-facing server.
>
> The CLI can be used to manage the data served by the server, and to start and stop the server.

Also - your data scheme seems to rely on machine fingerprints?  I'm not sure how that helps with docker containers.

That said (and since keygen-relay is not available, yet), do you have thoughts on the above?  I'm not sure how else to fingerprint a docker container.

**Reply by ezekg** [reply]
In the context of Relay, a "server" is Relay itself — a server run on a "customer's" offline network, distributing license files to the nodes on the network. The "app" here would be the "vendor's" application running on the nodes in the offline network, talking to Relay.

> Also - your data scheme seems to rely on machine fingerprints? I'm not sure how that helps with docker containers.

A fingerprint can be anything — a Docker container ID, an in-memory UUID, etc. It doesn't have to be something from the physical machine. Tons of people use Keygen's activation features to successfully license containers.

> That said (and since keygen-relay is not available, yet), do you have thoughts on the above? I'm not sure how else to fingerprint a docker container.

The above proposal is very complex, and it handles everything outside of Keygen, so I can't comment on how success that would be. I'd suggest simply checking out a machine file and transferring it to the offline machine. This is the "manual" flow that Relay is going to automate, since manually activating a machine is not always feasible especially for auto-scaling applications.

Here's an example: https://github.com/keygen-sh/air-gapped-activation-example

**Reply by al-dpopowich** [reply]
> A fingerprint can be anything — a Docker container ID, an in-memory UUID, etc. It doesn't have to be something from the physical machine. Tons of people use Keygen's activation features to successfully license containers.

But isn't the point of node-locking that part of the validation can confirm the license is for the machine it is running on?  Otherwise, I can just copy the machine-file to any other host, right?  I don't grok how an ephemeral source, like memory UUID or container ID can be a machine fingerprint.

(btw, thanks for the prompt replies!!!)

**Reply by ezekg** [reply]
If you copy the machine file to another container, the container ID or in-memory UUID would be different, and thus the machine file would fail to decrypt. I guess I'm not fully following your objection here, but feel free to expound. 🙂

**Reply by al-dpopowich** [reply]
You say my ideas are complex, yet you point me to that QR code example.  :wink:

I'm missing something.  Let's say I have an app that is distributed as docker images.  The user of the app is fully offline (air-gapped).  For them to be able to use the app they will need a license-file.  If I'm not worried about node-locking, I can:

1. Create a license. Checkout license and email the license-file and license-key.
2. Customer uses some internal means (e.g., usb-stick) to move the files to the target device.
3. By bind-mounting or other volume mgmt, the docker container will have access to the license-file and key.
4. Container is run.  App can see files and validates license. Trumpets call, the sun shines, life is good.

Can you provide a simple set of steps like this for node-locking in offline mode?  Bonus points if the app is shipped as a docker image.

Thank you so much!

**Reply by al-dpopowich** [reply]
> Here's an example: https://github.com/keygen-sh/air-gapped-activation-example

Btw, does this actually work?  When I use `ngrok` and go to the url, it asks to start my camera, then...nothing happens...just a page with my camera on.

**Reply by ezekg** [reply]
> You say my ideas are complex, yet you point me to that QR code example. 😉
> 
> I'm missing something. Let's say I have an app that is distributed as docker images. The user of the app is fully offline (air-gapped). For them to be able to use the app they will need a license-file. If I'm not worried about node-locking, I can:
> 
>     1. Create a license. Checkout license and email the license-file and license-key.
> 
>     2. Customer uses some internal means (e.g., usb-stick) to move the files to the target device.
> 
>     3. By bind-mounting or other volume mgmt, the docker container will have access to the license-file and key.
> 
>     4. Container is run.  App can see files and validates license. Trumpets call, the sun shines, life is good.
> 
> 
> Can you provide a simple set of steps like this for node-locking in offline mode? Bonus points if the app is shipped as a docker image.
> 
> Thank you so much!

This is right. Like I said, the same flow for node-locking especially for Docker images gets complex, because the flow always requires a human to facilitate the flow. That's why we're working on Relay, because the alternative _is_ complex because the problem is complex. Ideally, we get rid of the human altogether during the flow outside of onboarding/setup.

Like the air-gapped activation example shows, the current flow looks like this:

1. The application will prompt for the end-user's license key.
1. The application will "fingerprint" the air-gapped device.
1. The application will somehow display the fingerprint to the end-user.
1. The end-user will activate the license key + fingerprint out-of-band using an "activation portal" e.g. via a mobile phone.
1. The activation portal will perform a license validation API request for the license key, scoped to the fingerprint.
1. The activation portal will activate the fingerprint when required.
1. The activation portal will checkout a machine file, containing an encrypted license and machine object.
1. The end-user will download the machine file and transfer it to the client e.g. via USB drive.
1. The application will cryptographically verify the machine file using Ed25519.
1. The application will decrypt the machine file using AES-256-GCM.
2. The application will run normally.

What Relay will do is simplify the above to:

1. The vendor downloads Relay on the end-user's private network.
1. The vendor loads Relay with N signed/encrypted license files (allowing N nodes on the local network).
1. The vendor starts the Relay server on the local network.
1. The application will request a license file and key from Relay (e.g. from port `1234` on the local network).
1. The application will cryptographically verify the license file using Ed25519.
1. The application will decrypt the license file using AES-256-GCM.
1. The application will run normally.

In the case where there are N nodes already running, Relay will return an error and the application can fail to boot.

We get rid of the human (outside of setup), paving the way for a wider range of use cases.

> > Here's an example: https://github.com/keygen-sh/air-gapped-activation-example
> 
> Btw, does this actually work? When I use `ngrok` and go to the url, it asks to start my camera, then...nothing happens...just a page with my camera on.

You need to open the ngrok URL on a mobile device and scan the QR code.

**Reply by al-dpopowich** [reply]
Thank you so much for the "flow chart".  That helps.  For my customer (Acme in my above example), I am told that I may not assume an internet connection _at all_, only that files can be transferred, so a tool like Relay (which I love) may not even be possible for me in this situation.

In the end, given that "fingerprints" for docker images are at best ephemeral (e.g., after taking down a container and restarting one loses the CID, so you'd have to store the original CID in some obfuscated way), and given that no matter what, it's complex, I'll go with something like my original idea.  

Would this project be interested in my solution?  I'm imagining a single python script.

> You need to open the ngrok URL on a mobile device and scan the QR code.

I did.  When I open the URL, the app asks for permission to use my camera and when granted, that's all there is...a page showing my active camera.

**Reply by ezekg** [reply]
> Thank you so much for the "flow chart". That helps. For my customer (Acme in my above example), I am told that I may not assume an internet connection _at all_, only that files can be transferred, so a tool like Relay (which I love) may not even be possible for me in this situation.
> 
> In the end, given that "fingerprints" for docker images are at best ephemeral (e.g., after taking down a container and restarting one loses the CID, so you'd have to store the original CID in some obfuscated way), and given that no matter what, it's complex, I'll go with something like my original idea.
> 
> Would this project be interested in my solution? I'm imagining a single python script.

An internet connection is not needed for Relay. Relay serves license files on a local network, not via the internet. It's meant to be run in 100% air-gapped environments; it has no dependency on the internet — that's the entire point of Relay. If there's no local network, then Relay won't work, and you will need to do something like that aforementioned air-gapped activation flow, which requires human intervention.

Your solution requires trust on the client machine, so I'm not interested in it. It's impossible to fully trust the client, and any trust of the client in an offline environment, i.e. via obfuscation, will eventually be thwarted. In your solution, the client has the private key, so can generate unlimited activation requests in the worst-case, and I never assume best-case scenario i.r.t. licensing, only worst-case.

Your solution just adds complexity on top of the air-gapped flow I outlined using a machine file with a random fingerprint.

> I did. When I open the URL, the app asks for permission to use my camera and when granted, that's all there is...a page showing my active camera.

Use the camera to scan the QR code, per the instructions.

**Reply by al-dpopowich** [reply]
> An internet connection is not needed for Relay.

OK.  Read through your flow again. Got it now.  I was thinking the Relay server was on both networks, a gateway, as it were.  The name through me.

> Your solution requires trust on the client machine, so I'm not interested in it. It's impossible to fully trust the client, and any trust of the client in an offline environment, i.e. via obfuscation, will eventually be thwarted.

No disagreement here.  But that will also be true of Relay.  Any solution that is fully run on the client's host/network can be thwarted, no?  A little wireshark snooping and what's to stop a client from writing their own server to deliver false-positive results, allowing them to up their licenses?  It's a matter of how hard you make it for them.

That said, were Relay available now, I'd be taking it out for a test drive.

> Use the camera to scan the QR code, per the instructions.

I'm not sure what more I can say.  My camera turns on, but does nothing.  There are no instructions on that page, no buttons, no nothing, just a div with my camera view.  I see no way from there to scan the QR code.  I tried it with both chrome and firefox browsers on my android device. And if I use my phone's QR scanner, it gets a text file, but have no idea what to do with it from there.

**Reply by ezekg** [reply]
> Any solution that is fully run on the client's host/network can be thwarted, no?

Yes — anything run client-side can be thwarted. But Relay, using cryptography, does make it challenging. There are audit logs available in Relay if you do have a suspect client. These can be compared to your own application logs to see the whole picture.

> That said, were Relay available now, I'd be taking it out for a test drive.

It'll be in beta within the next couple weeks. I'll ping you.

**Reply by ezekg** [reply]
Keygen Relay is now in beta: https://keygen.sh/docs/relay/

GitHub: https://github.com/keygen-sh/keygen-relay

**Reply by al-dpopowich** [reply]
I'm finally getting back to this and need an understanding of the work-flow:

1. I create my app to communicate with a relay server to checkout a license.
2. I ship to my customer my application, relay, license file and necessary keys.
3. Customer, on their air-gapped network, adds the license to relay and starts the relay server.
4. Customer runs my app (configured via environment variables or some-such to know host:port of relay), checks out license, life is good.

But unless I'm missing something, there's still a huge level of trust here, right?  What's to stop the customer from installing multiple copies of relay, on multiple hosts, installing the license on each?

**Reply by ezekg** [reply]
Typically, Relay would be set up and configured by somebody on your team, not by the customer. The host and port should also be hard-coded into your application, the same as you would with a normal Keygen integration to `api.keygen.sh` — it should not a variable that your customer can point to wherever they want, the same way that your public key should be hard-coded into your application so that a bad actor can't just swap your public key out with their own by simply changing an environment variable.

With a hard-coded host and port, only a single Relay can be on the port you choose — typically for Relay that is `localhost:6349` — and that's enforced by the operating system. If your application only looks for Relay on port 6349, then any other Relay will be ignored.

**Reply by al-dpopowich** [reply]
If the point of relay is to offer an air-gapped solution, I don't see how I can configure and hard-code my app for an unknown network.  Unless I use, as you suggest, `localhost:6349`.  But that then just begs other questions. 

As I ask above, what prevents any customer from installing the app on multiple machines if the app's license is only checked against the relay instance on localhost?

[Here (in the Background section)](https://keygen.sh/docs/relay/) you suggest relay being the solution to "automated licensing of highly volatile nodes, e.g. autoscaling clusters of hundreds or even thousands of nodes", but I don't grok how that can work when locking apps down to localhost.

What would be wonderful is a full walk-through demo of how relay can be used for various licensing needs. As it stands now, it seems relay only guarantees payment of a single license, regardless of how much it is used by a customer.  And maybe that's all we can ever achieve in an offline mode.

**Reply by ezekg** [reply]
Thanks for the feedback. I only used `localhost` as an example for the simplest case, because anything more specific than that is going to largely depend on what the customer environment looks like. In most production environments you would advertise the Relay service to the network and other nodes would discover it, e.g. using mDNS at hostname `keygen-relay.local`:

1. Install Keygen Relay by downloading the binary onto a node in the local network.

2. Load Relay with the customer's licenses, e.g. if you're licensing 20 nodes, load 20 license file and key pairs.

3. Start the Relay server on the desired port, e.g. the default port `6349`:
   ```bash
   relay serve
   ```

4. Advertise the Relay service via mDNS to make it discoverable as `keygen-relay.local`:
   ```bash
   avahi-publish -a keygen-relay.local -R 192.168.1.100
   ```
   Replacing `192.168.1.100` with the node's IP.

5. Verify it’s discoverable and accessible from other nodes on the local network:
   ```bash
   curl http://keygen-relay.local:6349
   ```  
   The Relay server is now discoverable within the local network.

In this scenario, using mDNS, your application would hard-code and communicate with `keygen-relay.local` on port `6349`. mDNS will ensure that only a single Relay can advertise itself at the `keygen-relay.local` host. Of course, this can be bypassed with a reverse-proxy, but if you have that problem, I'd suggest regular audits into your customer environments. Right now, that's outside of the scope of the Relay project. But like I said, actual implementation will depend, because not everybody uses mDNS.

Networking setups vary widely, and some customers may use Consul, some internal DNS, some external DNS, some may use mDNS, some DNS-SD, etc. Setting up Relay for each customer will likely depend on what exactly that customer's local network looks like, and how they advertise and discover services, and what type of application you're licensing as well —

e.g. if you're distributing a containerized on-prem edition of a SaaS app, you may want to package Relay inside of the Helm chart or Docker compose configuration for the application, the same way you would e.g. Postgres, so all communication between nodes and Relay is setup automatically as part of your application's network and dependency stack.

Regardless, your application should know where to look depending on networking environment. Relay serves on a port, but the rest is left up to the integrator during the beta period. If your use-case doesn't fall into any of the aforementioned categories i.r.t. DNS, please share what Relay can do better here so that we can better facilitate what you need for a particular environment.

We'll work on improving the docs in this area for common setups. We're still learning here too, so we'll improve the docs as the beta progresses and we learn how our customers deploy and manage Relay inside of their customer environments.

Again, appreciate the feedback. Let me know if I can expound or clarify?

**Reply by al-dpopowich** [reply]
Complicating my problem are several factors:

- I am a contractor writing software for a company, let's call them Acme, and it is Acme that delivers the software to their customers.  Requiring hard-coding a hostname in the software I write, which in turn forces Acme to tell their customers they must install relay on a host with a specific DNS record isn't tenable.
- This is a complex docker-based application with a suite of services. Generating stable, deterministic machine IDs across container restarts across all the myriad ways containers can run is all but impossible.
- And the ultimate gotcha: Acme sells this software to conglomerates with multiple sites and one of their concerns is selling one license and it being used N times.  If the only guard against N copies is DNS I fear there's nothing relay can do for us in this particular use-case.

After discussions with Acme, we've decided we need a key-exchange solution similar to what I mentioned at the top of this thread.  It's not perfect. As we know, any offline solution is at best obfuscation, but at least it will prevent stealing licenses by simply copying the distro.

**Reply by ezekg** [reply]
In your scenario, Relay is designed to be shipped alongside your Docker-based application, as yet another service. It's not really special in this scenario. The same way, e.g. your application communicates with a single database service, it would communicate with a single Relay service on [the internal Docker network](https://docs.docker.com/engine/network/) (see also [Docker Compose networking](https://docs.docker.com/compose/how-tos/networking/) if you're using Compose).

```yaml
services:
  web:
    build: .
    ports:
      - "8000:8000"
  db:
    image: postgres
    ports:
      - "8001:5432"
  relay:
    image: keygen/relay
    ports:
      - "6349:6349"
```

The above Docker Compose config would expose Relay on `http://relay:6349` to all containers on the network, the same way Postgres would be available on `postgres://db:5432`. This would require no setup on Acme's end other than Acme, the vendor, loading Relay with *n* licenses for their customer (which you could automate for them), which would make it so up to *n* application nodes could be used concurrently, each claiming a license from Relay.

If this doesn't work for you, would you mind explaining why? I want to make sure I understand the blocker here.

**Reply by al-dpopowich** [reply]
I had prototyped this very solution.

What would prevent Ajax (Acme's customer) from taking the distribution and copying it over and over?  Every machine doing a `docker compose up` would see valid licenses, no?

**Reply by ezekg** [reply]
Like mentioned before, this scenario can be prevented the same way you would prevent multiple database instances from being run across multiple instances of `docker compose up` i.e. application instances. You would use a networking service like [Consul](https://developer.hashicorp.com/consul/docs/services/usage/define-services), run on a separate node, to register a single Relay service, also on a separate node, and all application nodes would communicate with Consul for service discovery. Consul would be responsible for maintaining the contract that a single Relay is registered and discoverable.

Is your application currently deployed using Compose?

Would a guide on using Consul and Relay be helpful?

**Reply by al-dpopowich** [reply]
If Acme, Inc. has a customer with multiple sites, one in Texas, another in Colorado, another in ..., etc, and they are air-gapped from each other, a single instance of relay is not possible.  In fact, a single instance of relay would not be possible.

**Reply by ezekg** [reply]
Thanks for the additional context. Relay is designed to license a single air-gapped environment, not multiple environments all air-gapped from each other. Relay needs some way to manage the air-gapped nodes collectively, and that's not possible when the environments can't communicate with a single Relay.

Relay would work if each environment was licensed separately, which is more typical, but that doesn't seem to be the case.

In this case, yes, Relay doesn't seem like a good fit and your solution seems fine.

Let me know if you need anything else. 👍

**Reply by al-dpopowich** [reply]
> Relay would work if each environment was licensed separately, which is more typical, but that doesn't seem to be the case.

Exactly.  But in our customer's case (or any case in which relay is used?), I don't see what can stop the end-user from copying relay + application to another network and "stealing" a license.

**Reply by ezekg** [reply]
I'm not really sure what you're looking for here, to be quite honest. The vendor is typically the one responsible for setting up Relay in the air-gapped environment, and if the customer is tampering with Relay or copying Relay in attempts to bypass their licensing restrictions, then internal audits into the customer environment should surface that behavior and it should be dealt with like any other breach of contract. I don't really think this is something Relay needs to be concerned with — in this scenario, the customer is a bad actor. Preventing this behavior from bad actors in a way that doesn't require internet access or an intermediary that has internet access is hard, or even impossible. Relay is meant to be an automated, stand-alone solution that is in sharp contrast to manual activation flows like you've outlined, because that flow doesn't work for containerized environments — esp. with autoscaling.

For example, assuming you're the vendor, an onsite tech of yours would install Consul and Relay into the customer's environment, load Relay with _n_ licenses, and your application instances would claim licenses from Relay via Consul, e.g. `http://relay.service.consul`. Preventing the customer from sabotaging the Relay is out of scope of our current project goals.

**Reply by al-dpopowich** [reply]
We definitely have differing opinions on what is too manual!  Having the customer send back a license-request to be processed, encrypted and signed (all with one go of running a script) is too manual, but having an onsite tech to configure and install is not? :grin:

**Reply by ezekg** [reply]
I think it's less of differing opinions and more that Relay is solving a different problem.

Your flow is more akin to what people already use machine files for — they make a 'license request', which at minimum consists of a fingerprint — packaged up however you see fit signed/encrypted/w/e. — and that fingerprint is then activated as a machine via the API using a 'portal', and then a machine file is checked out for the new machine. That machine file is then distributed and installed onto the air-gapped device. This flow requires a human to fulfill the 'license request', i.e. it's manual.

This manual fulfillment is problematic if you imagine GitHub wanting to license a 100-node k8s cluster of GitHub EE to a customer. The customer obviously doesn't want to use all 100 nodes if they don't need to, so these nodes need to autoscale. But going through that manual 'license request' flow every time k8s wants to autoscale doesn't make sense There are infinite manual tasks here: a 'license request' that needs to be fulfilled for every new node — and this is the problem Relay is trying to solve.

With Relay, you could set up Relay once with 100 licenses, inside of the k8s network, and the nodes could autoscale up to 100 nodes, all enforced by Relay. There a 1-time manual task here: setup Relay on-site and then that's it until the contract changes.

There's nothing wrong with your approach if it works for your use-case. People have been using machine files to successfully facilitate 'license requests' like that ever since we introduced them — in fact, that's why we designed machine files!

We'll work on improving our messaging with Relay. It's very early stages right now.

(By all means, feel free to implement and share your approach, since I think you originally asked if sharing it via a Python script would be helpful. I'm sure others would love to see it, including myself!)
